### PR TITLE
Enable ASAN for CoreCLR

### LINKED ISSUE #52131: Enable AddressSanitizer on Debug and Checked builds for both MSVC and Clang
AddressSanitizer is a tool that helps with common bugs like use-out-of-scope, buffer-overflow, use-after-free, etc. It is supported on GCC, Clang, and newer versions of MSVC.

Today, we use Run-Time Checks on MSVC to validate some of these scenarios, but we don't run any equivalent on non-MSVC builds. We have some infrastructure for running AddressSanitizer on Clang builds, but we've never enabled it for our builds (some partners have enabled it for theirs at various times).

By enabling AddressSanitizer for both our MSVC and Clang builds, we would ensure that we validate our code against common classes of bugs in C and C++.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Enable AddressSanitizer for CoreCLR Debug and Checked build configurations instead of MSVC RTC1 runtime checks.

Fixes https://github.com/dotnet/runtime/issues/52131

Contributes to https://github.com/dotnet/runtime/issues/7066 (gets the infra set up that we can actually explore this)

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
Looks like due to the architecture of ASAN, we likely won't be able to run libraries tests until we make the libraries test host use the live-built hosting bits.

**Comment by jkoritzinsky** [comment]
Reopening as a non-draft to keep the bot from closing it.

File: src/native/minipal/asansupport.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (janvorli) [inline src/native/minipal/asansupport.cpp:1]: I would suggest having only the code that's different for each platform here and have the rest out of the minipal. The intent of the minipal is to contain just Windows / Unix specific functionality separated into those two folders and all the methods being on the VMToOSInterface.

**Comment by AaronRobinsonMSFT** [comment]
Convert to draft?

**Comment by jkoritzinsky** [comment]
Yeah I can flip this back to draft. I just don't like when GitHub closes it due to it being a draft.

File: src/native/minipal/asansupport.cpp (hunk 1 above)
Comment (jkoritzinsky) [reply src/native/minipal/asansupport.cpp:1]: This is in the src/native minipal, not the CoreCLR minipal (the src/native minipal is just a set of free functions with a C-like interface)

**Comment by RobertHenry6bev** [comment]
Why was this closed?  I'm compiling the runtime with gcc-12 -fsanitize=undefined and uncovering dozens of problems.  I would clean up these errors first, then move on to tsan and asan.

**Comment by jkoritzinsky** [comment]
This is closed because I didn't have time to get back to getting this fully working. I'll try to get the branch cleaned up and a new PR opened.

**Pull request closed without merging**